Add basic SNMPv3 configuration and user handling

ubkr opened the pull request:
Adds basic SNMPv3 configuration to enable SNMPv3 on the device and configure parameters such as the listening port, EngineID, initial user and set its associated secrets.

ubkr committed:
Add SNMPv3 configuration and control.

Adds SNMPv3 configuration, user handling and control.

venki5685 commented on doc/Core.xml:
```diff
@@ -5352,6 +5362,592 @@ onvif://www.onvif.org/name/ARV-453
+  <section xml:id="s.snmpv3_extension">
+    <title>SNMPv3 Device Management Extension (Optional)</title>
+    <para>
+      This section defines an optional extension to the Device Management Service for supporting Simple Network Management Protocol version 3 (SNMPv3) as specified in <xref linkend="RFC3410"/> and related standards. Devices supporting this extension <emphasis role="bold">shall</emphasis> indicate this capability via the <literal>SNMPv3</literal> flag in the <literal>DeviceServiceCapabilities</lit ...
```
where SNMPv3 capability flag is added in DeviceServiceCapabilities in the PR?

kieran242 replied:
@ubkr @venki5685 there is an error in the table and the "SNMPv3" is not rendering properly in the table noted in my review comment. Looks to be missing a row section.

ubkr committed:
Add SNMPv3 capability flag to Network category.

kieran242 commented on doc/Core.xml:
```diff
@@ -5352,6 +5370,592 @@ onvif://www.onvif.org/name/ARV-453
+        <itemizedlist>
+          <listitem>
+            <para>
+              <emphasis role="bold">AES-128-CFB:</emphasis> As specified in <xref linkend="RFC3826"/> (OID: <literal>usmAesCfb128Protocol</literal>). This protocol is <emphasis role="bold">REQUIRED</emphasis>. The privacy key length is 16 octets.
```
@HansBusch do we need to also support these protocols "AES-128-CFB" in the Security Baseline doc? #571

kieran242 reviewed (commented):
@ubkr raised a few points and noted some of the Privacy protocols should be prob updated into the new Security Baseline document.

ubkr committed:
Corrected capabilities table rendering

ubkr committed:
Remove SNMPv3 dedicated user management operations

- Remove AddSNMPUser, DeleteSNMPUser, and GetSNMPUsers operations
- Enhance SetSNMPConfiguration to support initial user creation
- Remove SNMPUser and SNMPUserConfiguration complex types
- Retain security level, authentication, and privacy protocol types
- Update capability descriptions for service config + initial user setup
- Maintain GetSNMPConfiguration and SetSNMPConfiguration operations

ubkr committed:
Standardize SNMPv3 RFC reference style to match document conventions

- Remove xml:id attributes from SNMPv3 RFC entries (RFC 3410, 3411, 3414, 3826, 7860)
- Replace <xref linkend='RFCxxxx'/> with [RFC xxxx] bracket notation throughout SNMPv3 section
- Maintain datatracker.ietf.org URLs for modern RFC access
- Align SNMPv3 references with existing document RFC citation style

bsriramprasad committed:
schema validation errors

ubkr committed:
Fix SNMPv3 privacy protocol naming to align with RFC 3826

Rename privacy protocol enumerations from AES_*_CFB to CFB_AES_* to match
RFC 3826 naming convention (cipher mode first). Update OID references from
usmAesCfb*Protocol to usmAesCfb*PrivProtocol (correct "Priv" suffix).
Remove redundant SNMPv1/v2c statement.

The CFB_AES_128 pattern better represents cipher mode (CFB) followed by
algorithm (AES) and key size, consistent with RFC 3826 specifications.

ubkr committed:
Rename SNMPv3 to SNMP extension with scope clarification

Standardize terminology from "SNMPv3" to "SNMP" throughout specification
while clarifying that only SNMPv3 protocol version is supported. Update
section IDs, operation references, and capability flags for consistency.
Flatten SetSNMPConfiguration request structure to accept individual
parameters rather than nested types.

bsriramprasad committed:
spec edits for readability

bsriramprasad committed:
schema changes

ubkr commented:
Rebased to current development branch.

Some open questions to be adressed later:

* Which MIBs should be mandatory or recommended?
* Are the mentioned hash and encryption algoritms still valid in todays environment and are there others used by the industry and in IT envs?

venki5685 commented on doc/Core.xml:
```diff
@@ -5406,6 +5424,375 @@ onvif://www.onvif.org/name/ARV-453
+                      <entry>SecurityLevel</entry>
+                      <entry>tt:SNMPSecurityLevel</entry>
+                      <entry>optional</entry>
+                      <entry>Security level for the initial user (noAuthNoPriv, authNoPriv, authPriv). Required if InitialUserName is provided.</entry>
```
Device SecurityLevel supported values should be provided via SNMP Options for the client to choose and configure via SetSNMPConfiguration.

ubkr replied:
I've added a SNMPSecurityLevels capability

venki5685 commented on doc/Core.xml:
```diff
@@ -5406,6 +5424,375 @@ onvif://www.onvif.org/name/ARV-453
+                    </row>
+                    <row>
+                      <entry>AuthenticationProtocol</entry>
+                      <entry>tt:SNMPAuthenticationProtocol</entry>
```
Device AuthenticationProtocol supported values should be provided via SNMP Options for the client to choose and configure via SetSNMPConfiguration.

ubkr replied:
I've added a SNMPAuthenticationProtocols capability

venki5685 commented on doc/Core.xml:
```diff
@@ -5406,6 +5424,375 @@ onvif://www.onvif.org/name/ARV-453
+                      <entry>Authentication protocol. Required if SecurityLevel is authNoPriv or authPriv.</entry>
+                    </row>
+                    <row>
+                      <entry>AuthenticationKey</entry>
```
This parameter should be AuthPassword for Initial User, not AuthKey.

ubkr replied:
I changed it to AuthPassword as suggested

venki5685 commented on doc/Core.xml:
```diff
@@ -5406,6 +5424,375 @@ onvif://www.onvif.org/name/ARV-453
+                      <entry>Privacy protocol. Required if SecurityLevel is authPriv.</entry>
+                    </row>
+                    <row>
+                      <entry>PrivacyKey</entry>
```
This parameter should be PrivacyPassword for Initial User, not PrivacyKey.

ubkr replied:
I changed it to PrivacyPassword as suggested

venki5685 commented on doc/Core.xml:
```diff
@@ -5406,6 +5424,375 @@ onvif://www.onvif.org/name/ARV-453
+                    </row>
+                    <row>
+                      <entry>PrivacyProtocol</entry>
+                      <entry>tt:SNMPPrivacyProtocol</entry>
```
Device PrivacyProtocol supported values should be provided via SNMP Options for the client to choose and configure via SetSNMPConfiguration.

ubkr replied:
I've added a SNMPPrivacyProtocols capability

venki5685 commented on doc/Core.xml:
```diff
@@ -5406,6 +5424,375 @@ onvif://www.onvif.org/name/ARV-453
+                          port is maintained. Valid port range is typically 1-65535. </entry>
+                    </row>
+                    <row>
+                      <entry>InitialUserName</entry>
```
if the current SNMP V3 ONVIF spec is limited with only Initial User creation itself rather than SNMP multi user creation and it's management, the spec should explicitly clarify it's scope for snmp Initial User handling.

ubkr replied:
I tried to change the wording so that you can only set the parameters if SNMP is disabled and you make a SetSNMPConfiguration with all the parameters needed. if you want to change your configuration you need to disable SNMP and then redo the SetSNMPConfiguration with all parameters again.

venki5685 commented on wsdl/ver10/device/wsdl/devicemgmt.wsdl:
```diff
@@ -3677,13 +3768,24 @@ IN NO EVENT WILL THE CORPORATION OR ITS MEMBERS OR THEIR AFFILIATES BE LIABLE FO
+			<wsdl:output message="tds:GetSNMPConfigurationResponse"/>
+		</wsdl:operation>
+		<wsdl:operation name="SetSNMPConfiguration">
+			<wsdl:documentation>Enables or disables the SNMP service and allows configuration of global SNMP parameters such as engine ID and listening port, in accordance with Core.xml. Only SNMPv3 is supported. Only secure protocols as specified in Core.xml are accepted.</wsdl:documentation>
```
instead of saying Core.xml in the wsdl file, we can use ONVIF-Core-Specification which is published officially.

ubkr replied:
Refered to "core specifikation" as done in some other places in the document.

ubkr committed:
Enhance SNMP documentation with additional security protocol details and clarify parameter descriptions

ubkr committed:
Clarify SNMP security levels and protocols in documentation for improved understanding

ubkr committed:
Update SNMP documentation for clarity on core specification references

ubkr committed:
Update entry row span in Core.xml for accurate table representation

ubkr committed:
Align SNMP capability attributes between Core.xml and devicemgmt.wsdl

venki5685 commented on doc/Core.xml:
```diff
@@ -5406,6 +5449,393 @@ onvif://www.onvif.org/name/ARV-453
+      <para>
+        This section details the requirements for authentication and privacy protocols supported by the SNMP extension.
+        When an initial user is created or updated via <literal>SetSNMPConfiguration</literal>, the Device shall securely store localized keys and apply them according to the specified security level.
+        Authentication and privacy passwords shall be provided by the Client as plaintext strings. The Device shall derive the corresponding keys from these passwords using the password-to-key algorithm defined in [RFC 3414] §2.6, or as extended by [RFC 7860] §4 for SHA-2 based authentication protocols. The Device shall then localize the derived keys using its EngineID per [RFC 3414] Appendix A.1 ...
```
Since Authentication and privacy passwords shall be provided by the Client as plaintext strings in SetSNMPConfiguration API request, secure(i.e https) communication is recommended for this API.
